# PyCity Schools Analysis

In [585]:
# Dependencies and Setup
import pandas as pd
from pathlib import Path

# File to Load (Remember to Change These)
school_data_to_load = Path("../Resources/schools_complete.csv")
student_data_to_load = Path("../Resources/students_complete.csv")

# Read School and Student Data File and store into Pandas DataFrames
schools_df = pd.read_csv(school_data_to_load)
student_df = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
data_complete = pd.merge(student_df, schools_df, how="outer", on=["school_name", "school_name"])

Huang, Figueroa, Shelton, Hernandez, Griffin, Wilson, Cabrera, Bailey, Holden, Pena, Wright, Rodriguez, Johnson, Ford, Thomas

There are 15 schools - store in a variable

In [586]:
school_count = len(schools_df["school_name"])
student_count = len(student_df["Student ID"].unique())
print(student_count)

39170


data_complete["Student ID"].value_counts()
data_complete.describe()

this shows 39,170 students

length of the Student ID column is 39170   same for all columns

get enrollment count at each school from the size column

overall average math score stored in avg_math_score
overall average reading score stored in ave_reading_score

Creating a new column for pass or fail math score with 60 and higher being pass and lower being fail
and one for reading also

In [587]:

data_complete["math_pass"] = data_complete["math_score"].apply(lambda x:"pass" if x>=60 else "fail")

data_complete["reading_pass"] = data_complete["reading_score"].apply(lambda x:"pass" if x>=60 else "fail")

math_num_pass = data_complete["math_pass"].value_counts()
reading_num_pass = data_complete["reading_pass"].value_counts()

print(math_num_pass)
print(reading_num_pass)

data_complete.head()

pass    36211
fail     2959
Name: math_pass, dtype: int64
pass    39170
Name: reading_pass, dtype: int64


Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget math_pass  \
0             66          79          0  District  2917  1910635      pass   
1             94          61          0  District  2917  1910635      pass   
2             90          60          0  District  2917  1910635      pass   
3             67          58          0  District  2917  1910635      fail   
4             97          84          0  District  2917  1910635      pass   

  reading_pass  
0         pass  
1         pass  
2         pass  
3         pass  
4         pass

get percents passing
math_percent_pass
reading_percent_pass
overall_percent_pass

In [588]:
math_percent_pass = (math_num_pass / student_count)*100
reading_percent_pass = (reading_num_pass / student_count)*100

def get_status(row):
    if row["math_pass"] == "pass" and row["reading_pass"] == "pass":
        return "pass"
    
        

data_complete["overall pass"] = data_complete.apply(lambda row: get_status(row), axis=1)

overall_num_pass = data_complete["overall pass"].value_counts()

print(overall_num_pass)

overall_percent_pass = (overall_num_pass / student_count)*100

print(math_percent_pass)
print(reading_percent_pass)
print(overall_percent_pass)


pass    36211
Name: overall pass, dtype: int64
pass    92.445749
fail     7.554251
Name: math_pass, dtype: float64
pass    100.0
Name: reading_pass, dtype: float64
pass    92.445749
Name: overall pass, dtype: float64


Make a dataframe for each individual school

In [589]:
school_name_df = data_complete.set_index("school_name")

Huang_df = school_name_df.loc["Huang High School", :]

Huang_df.head()



Student ID       student_name gender grade  reading_score  \
school_name                                                                    
Huang High School           0       Paul Bradley      M   9th             66   
Huang High School           1       Victor Smith      M  12th             94   
Huang High School           2    Kevin Rodriguez      M  12th             90   
Huang High School           3  Dr. Richard Scott      M  12th             67   
Huang High School           4         Bonnie Ray      F   9th             97   

                   math_score  School ID      type  size   budget math_pass  \
school_name                                                                   
Huang High School          79          0  District  2917  1910635      pass   
Huang High School          61          0  District  2917  1910635      pass   
Huang High School          60          0  District  2917  1910635      pass   
Huang High School          58          0  District  2917  1910635      fail   
Huang High School          84          0  District  2917  1910635      pass   

                  reading_pass overall pass  
school_name                                  
Huang High School         pass         pass  
Huang High School         pass         pass  
Huang High School         pass         pass  
Huang High School         pass         None  
Huang High School         pass         pass

pass_lib=[math_pass_school1_df.iloc[1,0],math_pass_school1_df.iloc[2,0],math_pass_school1_df.iloc[4,0],math_pass_school1_df.iloc[6,0],
           math_pass_school1_df.iloc[7,0],math_pass_school1_df.iloc[9,0],math_pass_school1_df.iloc[10,0],math_pass_school1_df.iloc[12,0],
           math_pass_school1_df.iloc[14,0],math_pass_school1_df.iloc[15,0],math_pass_school1_df.iloc[17,0],math_pass_school1_df.iloc[18,0],
           math_pass_school1_df.iloc[19,0],math_pass_school1_df.iloc[20,0],math_pass_school1_df.iloc[21,0]]

In [590]:
grouped_school_totals = data_complete.groupby(["school_name","math_pass"])
math_pass_school_df = pd.DataFrame(grouped_school_totals[["math_pass"]].count())

grouped_school_totals2 = data_complete.groupby(["school_name","reading_pass"])
reading_pass_school_df = pd.DataFrame(grouped_school_totals2[["reading_pass"]].count())

math_pass_school1_df = math_pass_school_df.rename(columns = {"math_pass":"math_pass_or_fail"})
read_pass_school1_df = reading_pass_school_df.rename(columns = {"reading_pass":"reading_pass_or_fail"})
read_pass_school1_df


reading_pass_or_fail
school_name           reading_pass                      
Bailey High School    pass                          4976
Cabrera High School   pass                          1858
Figueroa High School  pass                          2949
Ford High School      pass                          2739
Griffin High School   pass                          1468
Hernandez High School pass                          4635
Holden High School    pass                           427
Huang High School     pass                          2917
Johnson High School   pass                          4761
Pena High School      pass                           962
Rodriguez High School pass                          3999
Shelton High School   pass                          1761
Thomas High School    pass                          1635
Wilson High School    pass                          2283
Wright High School    pass                          1800

In [591]:
pass_merge_df = pd.merge(schools_df, math_pass_school1_df, on= "school_name", how = "inner")
pass_merge2_df = pd.merge(pass_merge_df,read_pass_school1_df, on="school_name",how = "inner")


# drop the duplicate rows with the smaller number - the fails
pass_drop_df=pass_merge2_df.drop_duplicates(subset="school_name", keep="last")

renamed_pass_df = pass_drop_df.rename(columns={"math_pass_or_fail":"Num_pass_math","reading_pass_or_fail":"Num_pass_read"})

renamed_pass_df["% passing math"]=renamed_pass_df["Num_pass_math"]/renamed_pass_df["size"]*100
renamed_pass_df["% passing reading"]=renamed_pass_df["Num_pass_read"]/renamed_pass_df["size"]*100

#add column for per student budget
renamed_pass_df["Per student budget"]=renamed_pass_df["budget"]/renamed_pass_df["size"]
renamed_pass_df

School ID            school_name      type  size   budget  Num_pass_math  \
1           0      Huang High School  District  2917  1910635           2592   
3           1   Figueroa High School  District  2949  1884411           2608   
4           2    Shelton High School   Charter  1761  1056600           1761   
6           3  Hernandez High School  District  4635  3022020           4129   
7           4    Griffin High School   Charter  1468   917500           1468   
8           5     Wilson High School   Charter  2283  1319574           2283   
9           6    Cabrera High School   Charter  1858  1081356           1858   
11          7     Bailey High School  District  4976  3124928           4455   
12          8     Holden High School   Charter   427   248087            427   
13          9       Pena High School   Charter   962   585858            962   
14         10     Wright High School   Charter  1800  1049400           1800   
16         11  Rodriguez High School  District  3999  2547363           3541   
18         12    Johnson High School  District  4761  3094650           4246   
20         13       Ford High School  District  2739  1763916           2446   
21         14     Thomas High School   Charter  1635  1043130           1635   

    Num_pass_read  % passing math  % passing reading  Per student budget  
1            2917       88.858416              100.0               655.0  
3            2949       88.436758              100.0               639.0  
4            1761      100.000000              100.0               600.0  
6            4635       89.083064              100.0               652.0  
7            1468      100.000000              100.0               625.0  
8            2283      100.000000              100.0               578.0  
9            1858      100.000000              100.0               582.0  
11           4976       89.529743              100.0               628.0  
12            427      100.000000              100.0               581.0  
13            962      100.000000              100.0               609.0  
14           1800      100.000000              100.0               583.0  
16           3999       88.547137              100.0               637.0  
18           4761       89.182945              100.0               650.0  
20           2739       89.302665              100.0               644.0  
21           1635      100.000000              100.0               638.0

In [592]:
avg_school_scorem = data_complete.groupby(["school_name"])
avg_math_scores_df = pd.DataFrame(avg_school_scorem[["math_score"]].mean())

avg_read_scores_df =pd.DataFrame(avg_school_scorem[["reading_score"]].mean())

merge_scores_df = pd.merge(avg_math_scores_df,avg_read_scores_df,on = "school_name")

average_scores_df = merge_scores_df.rename(columns={"math_score":"Average math score","reading_score":"Average reading score"})
average_scores_df

Average math score  Average reading score
school_name                                                     
Bailey High School              77.048432              81.033963
Cabrera High School             83.061895              83.975780
Figueroa High School            76.711767              81.158020
Ford High School                77.102592              80.746258
Griffin High School             83.351499              83.816757
Hernandez High School           77.289752              80.934412
Holden High School              83.803279              83.814988
Huang High School               76.629414              81.182722
Johnson High School             77.072464              80.966394
Pena High School                83.839917              84.044699
Rodriguez High School           76.842711              80.744686
Shelton High School             83.359455              83.725724
Thomas High School              83.418349              83.848930
Wilson High School              83.274201              83.989488
Wright High School              83.682222              83.955000

In [593]:
final_merge_df = pd.merge(renamed_pass_df, average_scores_df, on= "school_name", how = "outer")
final_merge_df

School ID            school_name      type  size   budget  Num_pass_math  \
0           0      Huang High School  District  2917  1910635           2592   
1           1   Figueroa High School  District  2949  1884411           2608   
2           2    Shelton High School   Charter  1761  1056600           1761   
3           3  Hernandez High School  District  4635  3022020           4129   
4           4    Griffin High School   Charter  1468   917500           1468   
5           5     Wilson High School   Charter  2283  1319574           2283   
6           6    Cabrera High School   Charter  1858  1081356           1858   
7           7     Bailey High School  District  4976  3124928           4455   
8           8     Holden High School   Charter   427   248087            427   
9           9       Pena High School   Charter   962   585858            962   
10         10     Wright High School   Charter  1800  1049400           1800   
11         11  Rodriguez High School  District  3999  2547363           3541   
12         12    Johnson High School  District  4761  3094650           4246   
13         13       Ford High School  District  2739  1763916           2446   
14         14     Thomas High School   Charter  1635  1043130           1635   

    Num_pass_read  % passing math  % passing reading  Per student budget  \
0            2917       88.858416              100.0               655.0   
1            2949       88.436758              100.0               639.0   
2            1761      100.000000              100.0               600.0   
3            4635       89.083064              100.0               652.0   
4            1468      100.000000              100.0               625.0   
5            2283      100.000000              100.0               578.0   
6            1858      100.000000              100.0               582.0   
7            4976       89.529743              100.0               628.0   
8             427      100.000000              100.0               581.0   
9             962      100.000000              100.0               609.0   
10           1800      100.000000              100.0               583.0   
11           3999       88.547137              100.0               637.0   
12           4761       89.182945              100.0               650.0   
13           2739       89.302665              100.0               644.0   
14           1635      100.000000              100.0               638.0   

    Average math score  Average reading score  
0            76.629414              81.182722  
1            76.711767              81.158020  
2            83.359455              83.725724  
3            77.289752              80.934412  
4            83.351499              83.816757  
5            83.274201              83.989488  
6            83.061895              83.975780  
7            77.048432              81.033963  
8            83.803279              83.814988  
9            83.839917              84.044699  
10           83.682222              83.955000  
11           76.842711              80.744686  
12           77.072464              80.966394  
13           77.102592              80.746258  
14           83.418349              83.848930

Now cleaning the dataframe, renaming columnns, organizing, deleting unneeded columns
first - indexing by school name

In [594]:
full_school_df = final_merge_df.set_index("school_name")
del full_school_df["School ID"]
renamed_df = full_school_df.rename(columns = {"type":"School Type", "size":"Total Students", "budget":"Total school budget", "math_score":"Average math score",
                                       "reading_score":"Average reading score"})
organized_df=renamed_df[["School Type","Total Students", "Total school budget", "Per student budget", "Average math score", "Average reading score", "% passing math", "% passing reading"]]
organized_df

School Type  Total Students  Total school budget  \
school_name                                                              
Huang High School        District            2917              1910635   
Figueroa High School     District            2949              1884411   
Shelton High School       Charter            1761              1056600   
Hernandez High School    District            4635              3022020   
Griffin High School       Charter            1468               917500   
Wilson High School        Charter            2283              1319574   
Cabrera High School       Charter            1858              1081356   
Bailey High School       District            4976              3124928   
Holden High School        Charter             427               248087   
Pena High School          Charter             962               585858   
Wright High School        Charter            1800              1049400   
Rodriguez High School    District            3999              2547363   
Johnson High School      District            4761              3094650   
Ford High School         District            2739              1763916   
Thomas High School        Charter            1635              1043130   

                       Per student budget  Average math score  \
school_name                                                     
Huang High School                   655.0           76.629414   
Figueroa High School                639.0           76.711767   
Shelton High School                 600.0           83.359455   
Hernandez High School               652.0           77.289752   
Griffin High School                 625.0           83.351499   
Wilson High School                  578.0           83.274201   
Cabrera High School                 582.0           83.061895   
Bailey High School                  628.0           77.048432   
Holden High School                  581.0           83.803279   
Pena High School                    609.0           83.839917   
Wright High School                  583.0           83.682222   
Rodriguez High School               637.0           76.842711   
Johnson High School                 650.0           77.072464   
Ford High School                    644.0           77.102592   
Thomas High School                  638.0           83.418349   

                       Average reading score  % passing math  \
school_name                                                    
Huang High School                  81.182722       88.858416   
Figueroa High School               81.158020       88.436758   
Shelton High School                83.725724      100.000000   
Hernandez High School              80.934412       89.083064   
Griffin High School                83.816757      100.000000   
Wilson High School                 83.989488      100.000000   
Cabrera High School                83.975780      100.000000   
Bailey High School                 81.033963       89.529743   
Holden High School                 83.814988      100.000000   
Pena High School                   84.044699      100.000000   
Wright High School                 83.955000      100.000000   
Rodriguez High School              80.744686       88.547137   
Johnson High School                80.966394       89.182945   
Ford High School                   80.746258       89.302665   
Thomas High School                 83.848930      100.000000   

                       % passing reading  
school_name                               
Huang High School                  100.0  
Figueroa High School               100.0  
Shelton High School                100.0  
Hernandez High School              100.0  
Griffin High School                100.0  
Wilson High School                 100.0  
Cabrera High School                100.0  
Bailey High School                 100.0  
Holden High School                 100.0  
Pena High School                   100.0  
Wright High School                 100.0  
Rodriguez

formatting values in this data frame
code is commented out because numerical values are needed for sorting and grouping later.  Copy this code to the end and adjust to format final results data frames.

def format_int():
    return "{:,}".format()

def format_per():
    return "{:.1f}%".format()

def format_budg():
    return "${:,}".format()

def format_psbudget():
    return "${.0f}".format()

organized_df.loc[:,"Total Students"] = organized_df.loc[:,"Total Students"].apply(format_int)


In [595]:
organized_df["% overall passing"]=organized_df["% passing math"]
organized_df.head()

C:\Users\mrkol\AppData\Local\Temp\ipykernel_4144\1385722768.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  organized_df["% overall passing"]=organized_df["% passing math"]


School Type  Total Students  Total school budget  \
school_name                                                              
Huang High School        District            2917              1910635   
Figueroa High School     District            2949              1884411   
Shelton High School       Charter            1761              1056600   
Hernandez High School    District            4635              3022020   
Griffin High School       Charter            1468               917500   

                       Per student budget  Average math score  \
school_name                                                     
Huang High School                   655.0           76.629414   
Figueroa High School                639.0           76.711767   
Shelton High School                 600.0           83.359455   
Hernandez High School               652.0           77.289752   
Griffin High School                 625.0           83.351499   

                       Average reading score  % passing math  \
school_name                                                    
Huang High School                  81.182722       88.858416   
Figueroa High School               81.158020       88.436758   
Shelton High School                83.725724      100.000000   
Hernandez High School              80.934412       89.083064   
Griffin High School                83.816757      100.000000   

                       % passing reading  % overall passing  
school_name                                                  
Huang High School                  100.0          88.858416  
Figueroa High School               100.0          88.436758  
Shelton High School                100.0         100.000000  
Hernandez High School              100.0          89.083064  
Griffin High School                100.0         100.000000

Now sort the school information.  % overall passing must not be mapped yet, so that it can be used to sort numerically

In [596]:
top_schools = organized_df.sort_values(by=["% overall passing","Total Students"], ascending=False).head()
top_schools["Total Students"] = top_schools["Total Students"].map("{:,}".format)
top_schools["% passing math"] = top_schools["% passing math"].map("{:.1f}%".format)
top_schools["% overall passing"] = top_schools["% overall passing"].map("{:.1f}%".format)
top_schools["Total school budget"] = top_schools["Total school budget"].map("${:,}".format)
top_schools["% passing reading"] = top_schools["% passing reading"].map("{:.1f}%".format)
top_schools["Average reading score"] = top_schools["Average reading score"].map("{:.1f}".format)
top_schools["Average math score"] = top_schools["Average math score"].map("{:.1f}".format)
top_schools

School Type Total Students Total school budget  \
school_name                                                          
Wilson High School      Charter          2,283          $1,319,574   
Cabrera High School     Charter          1,858          $1,081,356   
Wright High School      Charter          1,800          $1,049,400   
Shelton High School     Charter          1,761          $1,056,600   
Thomas High School      Charter          1,635          $1,043,130   

                     Per student budget Average math score  \
school_name                                                  
Wilson High School                578.0               83.3   
Cabrera High School               582.0               83.1   
Wright High School                583.0               83.7   
Shelton High School               600.0               83.4   
Thomas High School                638.0               83.4   

                    Average reading score % passing math % passing reading  \
school_name                                                                  
Wilson High School                   84.0         100.0%            100.0%   
Cabrera High School                  84.0         100.0%            100.0%   
Wright High School                   84.0         100.0%            100.0%   
Shelton High School                  83.7         100.0%            100.0%   
Thomas High School                   83.8         100.0%            100.0%   

                    % overall passing  
school_name                            
Wilson High School             100.0%  
Cabrera High School            100.0%  
Wright High School             100.0%  
Shelton High School            100.0%  
Thomas High School             100.0%

the top schools are organized by % overall passing, and then by Total Students in descending order, as larger schools have lower passing rates in general.  It is most impressive to achieve 100% passing for a large school

In [597]:
bottom_schools = organized_df.sort_values(by=["% overall passing"], ascending=True).head()
bottom_schools["Total Students"] = bottom_schools["Total Students"].map("{:,}".format)
bottom_schools["% passing math"] = bottom_schools["% passing math"].map("{:.1f}%".format)
bottom_schools["% overall passing"] = bottom_schools["% overall passing"].map("{:.1f}%".format)
bottom_schools["Total school budget"] = bottom_schools["Total school budget"].map("${:,}".format)
bottom_schools["% passing reading"] = bottom_schools["% passing reading"].map("{:.1f}%".format)
bottom_schools["Average reading score"] = bottom_schools["Average reading score"].map("{:.1f}".format)
bottom_schools["Average math score"] = bottom_schools["Average math score"].map("{:.1f}".format)
bottom_schools

School Type Total Students Total school budget  \
school_name                                                            
Figueroa High School     District          2,949          $1,884,411   
Rodriguez High School    District          3,999          $2,547,363   
Huang High School        District          2,917          $1,910,635   
Hernandez High School    District          4,635          $3,022,020   
Johnson High School      District          4,761          $3,094,650   

                       Per student budget Average math score  \
school_name                                                    
Figueroa High School                639.0               76.7   
Rodriguez High School               637.0               76.8   
Huang High School                   655.0               76.6   
Hernandez High School               652.0               77.3   
Johnson High School                 650.0               77.1   

                      Average reading score % passing math % passing reading  \
school_name                                                                    
Figueroa High School                   81.2          88.4%            100.0%   
Rodriguez High School                  80.7          88.5%            100.0%   
Huang High School                      81.2          88.9%            100.0%   
Hernandez High School                  80.9          89.1%            100.0%   
Johnson High School                    81.0          89.2%            100.0%   

                      % overall passing  
school_name                              
Figueroa High School              88.4%  
Rodriguez High School             88.5%  
Huang High School                 88.9%  
Hernandez High School             89.1%  
Johnson High School               89.2%

## Math and Reading Scores by Grade

they go in order of 10th, 11th, 12th, 9th.  

In [614]:
ninth_graders = data_complete[(data_complete["grade"] == "9th")]
tenth_graders = data_complete[(data_complete["grade"] == "10th")]
eleventh_graders = data_complete[(data_complete["grade"] == "11th")]
twelfth_graders = data_complete[(data_complete["grade"] == "12th")]

ninth_grade_math_scores = pd.DataFrame(ninth_graders.groupby(["school_name","grade"]).mean())
tenth_grade_math_scores = pd.DataFrame(tenth_graders.groupby(["school_name","grade"]).mean())
eleventh_grade_math_scores = pd.DataFrame(eleventh_graders.groupby(["school_name","grade"]).mean())
twelfth_grade_math_scores = pd.DataFrame(twelfth_graders.groupby(["school_name","grade"]).mean())

columns_to_drop = ["reading_score","Student ID","School ID", "size", "budget"]

ninth_grade_math_scores = ninth_grade_math_scores.drop(columns_to_drop, axis=1)
ninth_grade = ninth_grade_math_scores.rename(columns = {"math_score":"9th"})

tenth_grade_math_scores = tenth_grade_math_scores.drop(columns_to_drop, axis=1)
tenth_grade = tenth_grade_math_scores.rename(columns = {"math_score":"10th"})

eleventh_grade_math_scores = eleventh_grade_math_scores.drop(columns_to_drop, axis=1)
eleventh_grade = eleventh_grade_math_scores.rename(columns = {"math_score":"11th"})

twelfth_grade_math_scores = twelfth_grade_math_scores.drop(columns_to_drop, axis=1)
twelfth_grade = twelfth_grade_math_scores.rename(columns = {"math_score":"12th"})

ninth_grade = ninth_grade.droplevel("grade")
tenth_grade = tenth_grade.droplevel("grade")
eleventh_grade = eleventh_grade.droplevel("grade")
twelfth_grade = twelfth_grade.droplevel("grade")

math_merge1 = pd.merge(ninth_grade, tenth_grade, on="school_name")
math_merge2 = pd.merge(math_merge1,eleventh_grade, on="school_name")
math_scores_by_grade = pd.merge(math_merge2, twelfth_grade, on="school_name")
math_scores_by_grade

9th       10th       11th       12th
school_name                                                      
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

In [615]:
ninth_grade_reading_scores = pd.DataFrame(ninth_graders.groupby(["school_name","grade"]).mean())
tenth_grade_reading_scores = pd.DataFrame(tenth_graders.groupby(["school_name","grade"]).mean())
eleventh_grade_reading_scores = pd.DataFrame(eleventh_graders.groupby(["school_name","grade"]).mean())
twelfth_grade_reading_scores = pd.DataFrame(twelfth_graders.groupby(["school_name","grade"]).mean())

columns_to_drop = ["math_score","Student ID","School ID", "size", "budget"]

ninth_grade_reading_scores = ninth_grade_reading_scores.drop(columns_to_drop, axis=1)
ninth_grader = ninth_grade_reading_scores.rename(columns = {"reading_score":"9th"})

tenth_grade_reading_scores = tenth_grade_reading_scores.drop(columns_to_drop, axis=1)
tenth_grader = tenth_grade_reading_scores.rename(columns = {"reading_score":"10th"})

eleventh_grade_reading_scores = eleventh_grade_reading_scores.drop(columns_to_drop, axis=1)
eleventh_grader = eleventh_grade_reading_scores.rename(columns = {"reading_score":"11th"})

twelfth_grade_reading_scores = twelfth_grade_reading_scores.drop(columns_to_drop, axis=1)
twelfth_grader = twelfth_grade_reading_scores.rename(columns = {"reading_score":"12th"})

ninth_grader = ninth_grader.droplevel("grade")
tenth_grader = tenth_grader.droplevel("grade")
eleventh_grader = eleventh_grader.droplevel("grade")
twelfth_grader = twelfth_grader.droplevel("grade")

reading_merge1 = pd.merge(ninth_grader, tenth_grader, on="school_name")
reading_merge2 = pd.merge(reading_merge1,eleventh_grader, on="school_name")
reading_scores_by_grade = pd.merge(reading_merge2, twelfth_grader, on="school_name")
reading_scores_by_grade

9th       10th       11th       12th
school_name                                                      
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

## Scores by School Spending

In [ ]:
spending_bins = [0,585,630,645,680]
labels = ["<$585", "$585-$630", "$630-$645", "$645-$680"]

organized_df["Spending Ranges(Per Student)"]=pd.cut(organized_df["Per student budget"],spending_bins, labels=labels)

spending_math_scores = organized_df.groupby(["Spending Ranges(Per Student)"])["Average math score"].mean()
spending_reading_scores = organized_df.groupby(["Spending Ranges(Per Student)"])["Average reading score"].mean()
spending_passing_math = organized_df.groupby(["Spending Ranges(Per Student)"])["% passing math"].mean()
spending_passing_reading = organized_df.groupby(["Spending Ranges(Per Student)"])["% passing reading"].mean()
overall_passing_spending = organized_df.groupby(["Spending Ranges(Per Student)"])["% overall passing"].mean()





In [ ]:
spend_merge_1 = pd.merge(spending_math_scores, spending_reading_scores, on="Spending Ranges(Per Student)")
spend_merge_2 = pd.merge(spend_merge_1, spending_passing_math, on="Spending Ranges(Per Student)")
spend_merge_3 = pd.merge(spend_merge_2, spending_passing_reading, on="Spending Ranges(Per Student)")
spending_summary = pd.merge(spend_merge_3,overall_passing_spending, on = "Spending Ranges(Per Student)")


map(format) numerical values in this completed data frame

In [ ]:
spending_summary["% passing reading"] = spending_summary["% passing reading"].map("{:.1f}%".format)
spending_summary["% passing math"] = spending_summary["% passing math"].map("{:.1f}%".format)
spending_summary["% overall passing"] = spending_summary["% overall passing"].map("{:.1f}%".format)
spending_summary["Average reading score"] = spending_summary["Average reading score"].map("{:.1f}".format)
spending_summary["Average math score"] = spending_summary["Average math score"].map("{:.1f}".format)

spending_summary

## Scores by School Size

In [ ]:
size_bins = [0,1000,2000,5000]

labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

organized_df["Total Students Ranges"]=pd.cut(organized_df["Total Students"],size_bins, labels=labels)

size_math_scores = organized_df.groupby(["Total Students Ranges"])["Average math score"].mean()
size_reading_scores = organized_df.groupby(["Total Students Ranges"])["Average reading score"].mean()
size_passing_math = organized_df.groupby(["Total Students Ranges"])["% passing math"].mean()
size_passing_reading = organized_df.groupby(["Total Students Ranges"])["% passing reading"].mean()
size_passing_overall = organized_df.groupby(["Total Students Ranges"])["% overall passing"].mean()
size_student_budget = organized_df.groupby(["Total Students Ranges"])["Per student budget"].mean()

In [ ]:
size_merge_1 = pd.merge(size_math_scores, size_reading_scores, on="Total Students Ranges")
size_merge_2 = pd.merge(size_merge_1, size_passing_math, on="Total Students Ranges")
size_merge_3 = pd.merge(size_merge_2, size_passing_reading, on="Total Students Ranges")
size_merge_4 = pd.merge(size_merge_3,size_student_budget, on = "Total Students Ranges")
size_summary = pd.merge(size_merge_4,size_passing_overall, on = "Total Students Ranges")

In [ ]:
size_summary["% passing reading"] = size_summary["% passing reading"].map("{:.1f}%".format)
size_summary["% passing math"] = size_summary["% passing math"].map("{:.1f}%".format)
size_summary["% overall passing"] = size_summary["% overall passing"].map("{:.1f}%".format)
size_summary["Average reading score"] = size_summary["Average reading score"].map("{:.1f}".format)
size_summary["Average math score"] = size_summary["Average math score"].map("{:.1f}".format)
size_summary["Per student budget"] = size_summary["Per student budget"].map("${:.2f}".format)

## Scores by School Type

In [ ]:
type_math_scores = organized_df.groupby(["School Type"])["Average math score"].mean()
type_reading_scores = organized_df.groupby(["School Type"])["Average reading score"].mean()
type_passing_math = organized_df.groupby(["School Type"])["% passing math"].mean()
type_passing_reading = organized_df.groupby(["School Type"])["% passing reading"].mean()
type_passing_overall = organized_df.groupby(["School Type"])["% overall passing"].mean()
type_student_budget = organized_df.groupby(["School Type"])["Per student budget"].mean()

In [ ]:
type_merge_1 = pd.merge(type_math_scores, type_reading_scores, on="School Type")
type_merge_2 = pd.merge(type_merge_1, type_passing_math, on="School Type")
type_merge_3 = pd.merge(type_merge_2, type_passing_reading, on="School Type")
type_merge_4 = pd.merge(type_merge_3, type_student_budget, on= "School Type")
type_summary = pd.merge(type_merge_4,type_passing_overall, on = "School Type")

In [ ]:
type_summary["% passing reading"] = type_summary["% passing reading"].map("{:.1f}%".format)
type_summary["% passing math"] = type_summary["% passing math"].map("{:.1f}%".format)
type_summary["% overall passing"] = type_summary["% overall passing"].map("{:.1f}%".format)
type_summary["Average reading score"] = type_summary["Average reading score"].map("{:.1f}".format)
type_summary["Average math score"] = type_summary["Average math score"].map("{:.1f}".format)
type_summary["Per student budget"] = type_summary["Per student budget"].map("${:.2f}".format)

In [ ]:
organized_df["Total Students"] = organized_df["Total Students"].map("{:,}".format)
organized_df["Total school budget"] = organized_df["Total school budget"].map("${:,}".format)
organized_df["Per student budget"] = organized_df["Per student budget"].map("${:.0f}".format)
organized_df["Average math score"] = organized_df["Average math score"].map("{:.1f}".format)
organized_df["Average reading score"] = organized_df["Average reading score"].map("{:.1f}".format)
organized_df["% passing reading"] = organized_df["% passing reading"].map("{:.1f}%".format)